src/coreclr/vm/riscv64/asmhelpers.S

```diff
@@ -1,7 +1,967 @@
+    .quad 0
+wbs_ephemeral_high:
+    .quad 0
+wbs_lowest_address:
+    .quad 0
+wbs_highest_address:
+    .quad 0
+WRITE_BARRIER_END JIT_WriteBarrier_Table
+
+// ------------------------------------------------------------------
+// End of the writeable code region
+LEAF_ENTRY JIT_PatchedCodeLast, _TEXT
+    ret
+LEAF_END JIT_PatchedCodeLast, _TEXT
+
+
+//
+// If a preserved register were pushed onto the stack between
+// the managed caller and the H_M_F, ptrS0_S8 will point to its
+// location on the stack and it would have been updated on the
+// stack by the GC already and it will be popped back into the
+// appropriate register when the appropriate epilog is run.
+//
+// Otherwise, the register is preserved across all the code
+// in this HCALL or FCALL, so we need to update those registers
+// here because the GC will have updated our copies in the
+// frame.
+//
+// So, if ptrS0_S8 points into the MachState, we need to update
+// the register here.  That's what this macro does.
+//
+.macro RestoreRegMS idx, reg
+    // Incoming:
+    //
+    // a0 = address of MachState
+    //
+    // idx: Index of the callee register
+    // s0/fp: 0, s1: 1, s3-s11: 4-11, gp: 12 tp: 13
+    //
+    // reg: Register name (e.g. s0, s1, etc)
+    //
+    // Get the address of the specified captured register from machine state
+    addi  a2, a0, (MachState__captureCalleeSavedRegisters + (\idx * 8))
+
+    //// Get the content of specified preserved register pointer from machine state
+    ld  a3, (MachState__ptrCalleeSavedRegisters + (\idx * 8))(a0)
+
+    bne  a2, a3, LOCAL_LABEL(NoRestore_\reg)
+
+    ld  \reg, 0(a2)
+LOCAL_LABEL(NoRestore_\reg):
+
+.endm
+
+NESTED_ENTRY ThePreStub, _TEXT, NoHandler
+    PROLOG_WITH_TRANSITION_BLOCK
+    .cfi_adjust_cfa_offset __PWTB_StackAlloc
+
+    addi  a1, METHODDESC_REGISTER, 0 // pMethodDesc
+
+    addi  a0, sp, __PWTB_TransitionBlock        // pTransitionBlock
+    call  PreStubWorker
+    addi  t4, a0, 0
+
+    EPILOG_WITH_TRANSITION_BLOCK_TAILCALL
+    .cfi_adjust_cfa_offset -__PWTB_StackAlloc
+    EPILOG_BRANCH_REG  t4
+NESTED_END ThePreStub, _TEXT
+
+// ------------------------------------------------------------------\
+
+// EXTERN_C int __fastcall HelperMethodFrameRestoreState(
+// INDEBUG_COMMA(HelperMethodFrame *pFrame)
+// MachState *pState
+// )
+LEAF_ENTRY HelperMethodFrameRestoreState, _TEXT
+#ifdef _DEBUG
+    addi  a0, a1, 0
+#endif
+
+    // If machine state is invalid, then simply exit
+    lw  a1, MachState__isValid(a0)
+    beq  a1, zero, LOCAL_LABEL(Done)
+
+    // manually assign index
+    // s0/fp: 0, s1: 1, s3-s11: 4-11, gp: 12 tp: 13
+    RestoreRegMS  0, s0 
+    RestoreRegMS  1, s1
+    RestoreRegMS  2, s2
+    RestoreRegMS  3, s3
+    RestoreRegMS  4, s4
+    RestoreRegMS  5, s5
+    RestoreRegMS  6, s6
+    RestoreRegMS  7, s7
+    RestoreRegMS  8, s8
+    RestoreRegMS  9, s9
+    RestoreRegMS  10, s10
+    RestoreRegMS  11, s11
+    RestoreRegMS  12, tp
+    RestoreRegMS  13, gp
```

Comment:
As said in `stubs.cpp` and in comment above `gp` is 12, `sp` is 13.

Reply:
Thank you. Fixed